### PR TITLE
Cleaning up test_corpus code

### PR DESCRIPTION
**Pull request by aisamanra**
### Motivation
The `test_corpus.cc` file has a lot of repetition and it makes following the logic of what's actually happening there rather opaque. In the interest of making it clearer, easier to follow, and less repetitive, this PR introduces a common abstraction that allows us to factor out the repetition and make the test structure clearer.

Note that this will be somewhat slower to compile than before, as it's using templates for closure types in places. I'd argue the change is worth it (especially as it's a test) but it's worth noting.

### Test plan

I manually went through and broke some tests to make sure that this still has the same semantics, but I don't know if it's worth including test-tests for this as well.

### CODE REVIEW

File: test/test_corpus.cc [hunk 1]
```diff
@@ -132,6 +132,56 @@ unique_ptr<Diagnostic> errorToDiagnostic(const core::GlobalState &gs, const core
     return make_unique<Diagnostic>(Range::fromLoc(gs, error.loc), error.header);
 }
 
+class ErrorHandler {
+    vector<unique_ptr<core::Error>> &errors;
```
Comment (jvilk-stripe) [inline test/test_corpus.cc]: Is there a reason why `ErrorHandler` doesn't own `errors`? Seems like a footgun to store a reference to it.

File: test/test_corpus.cc [hunk 2]
```diff
@@ -132,6 +132,56 @@ unique_ptr<Diagnostic> errorToDiagnostic(const core::GlobalState &gs, const core
...
+    Expectations &test;
```
Comment (jvilk-stripe) [inline test/test_corpus.cc]: (A reference to expectations makes sense, as it's a global entity whose lifetime is linked w/ the test.)

File: test/test_corpus.cc [hunk 3]
```diff
@@ -132,6 +132,56 @@ unique_ptr<Diagnostic> errorToDiagnostic(const core::GlobalState &gs, const core
...
+    shared_ptr<core::ErrorQueue> errorQueue;
+
+public:
+    map<string, string> got;
+
+    ErrorHandler(vector<unique_ptr<core::Error>> &errors, Expectations &test, shared_ptr<core::ErrorQueue> errorQueue)
```
Comment (jvilk-stripe) [inline test/test_corpus.cc]: Nit: You can make the `errorQueue` param a reference; I think as-is the code will cause two atomic increments of the `shared_ptr` reference counter -- one to create the argument, and one to assign to the class field.

File: test/test_corpus.cc [hunk 4]
```diff
@@ -132,6 +132,56 @@ unique_ptr<Diagnostic> errorToDiagnostic(const core::GlobalState &gs, const core
     return make_unique<Diagnostic>(Range::fromLoc(gs, error.loc), error.header);
 }
 
+class ErrorHandler {
+    vector<unique_ptr<core::Error>> &errors;
+    Expectations &test;
+    shared_ptr<core::ErrorQueue> errorQueue;
+
+public:
+    map<string, string> got;
```
Comment (jvilk-stripe) [inline test/test_corpus.cc]: Could this be an `UnorderedMap`? (`map` is actually an ordered tree.)

File: test/test_corpus.cc [hunk 5]
```diff
@@ -132,6 +132,56 @@ unique_ptr<Diagnostic> errorToDiagnostic(const core::GlobalState &gs, const core
...
+
+    ErrorHandler(vector<unique_ptr<core::Error>> &errors, Expectations &test, shared_ptr<core::ErrorQueue> errorQueue)
+        : errors(errors), test(test), errorQueue(errorQueue){};
+
+    void checkExpectations(string expectationType, std::function<string()> mkExp, bool addNewline = true) {
```
Comment (jvilk-stripe) [inline test/test_corpus.cc]: Nit: Can `expectationType` be a `string_view` (or does that not work with `got.contains`)? I believe it should work with the `UnorderedMap`...

File: test/test_corpus.cc [hunk 6]
```diff
@@ -132,6 +132,56 @@ unique_ptr<Diagnostic> errorToDiagnostic(const core::GlobalState &gs, const core
...
+        }
+    }
+
+    void drainErrors() {
+        auto newErrors = errorQueue->drainAllErrors();
+        errors.insert(errors.end(), make_move_iterator(newErrors.begin()), make_move_iterator(newErrors.end()));
+    }
+
+    void clear() {
+        got.clear();
```
Comment (jvilk-stripe) [inline test/test_corpus.cc]: Should this clear `errors` too?

File: test/test_corpus.cc [hunk 7]
```diff
@@ -132,6 +132,56 @@ unique_ptr<Diagnostic> errorToDiagnostic(const core::GlobalState &gs, const core
...
+            got[expectationType].append(mkExp());
+            if (addNewline) {
+                got[expectationType].append("\n");
+            }
+            auto newErrors = errorQueue->drainAllErrors();
+            errors.insert(errors.end(), make_move_iterator(newErrors.begin()), make_move_iterator(newErrors.end()));
+        }
+    }
+
+    void verifyAll(string prefix = "") {
```
Comment (jvilk-stripe) [inline test/test_corpus.cc]: Ditto: can this be a `string_view`?

File: test/test_corpus.cc [hunk 8]
```diff
@@ -132,6 +132,56 @@ unique_ptr<Diagnostic> errorToDiagnostic(const core::GlobalState &gs, const core
...
+        for (auto &gotPhase : got) {
+            auto expectation = test.expectations.find(gotPhase.first);
+            ASSERT_TRUE(expectation != test.expectations.end())
+                << prefix << "missing expectation for " << gotPhase.first;
+            ASSERT_TRUE(expectation->second.size() == 1)
+                << prefix << "found unexpected multiple expectations of type " << gotPhase.first;
+
+            auto checker = test.folder + expectation->second.begin()->second;
+            auto expect = FileOps::read(checker.c_str());
```
Comment (jvilk-stripe) [inline test/test_corpus.cc]: I *think* the `c_str` call is no longer required.

File: test/test_corpus.cc (hunk 5 above)
Comment (jvilk-stripe) [reply test/test_corpus.cc]: Also, `checkExpectations` is a misnomer; this is actually more like `storeActual`?

File: test/test_corpus.cc (hunk 7 above)
Comment (jvilk-stripe) [reply test/test_corpus.cc]: Maybe *this* should be called "checkExpectations"...

File: test/test_corpus.cc [hunk 9]
```diff
@@ -132,6 +132,56 @@ unique_ptr<Diagnostic> errorToDiagnostic(const core::GlobalState &gs, const core
     return make_unique<Diagnostic>(Range::fromLoc(gs, error.loc), error.header);
 }
 
+class ErrorHandler {
```
Comment (jvilk-stripe) [inline test/test_corpus.cc]: Any strong feelings on name? Should this be something like "ExpectationChecker"?

Comment (aisamanra) [reply test/test_corpus.cc]: I ended up going with the Solomonic choice of `ExpectationHandler`.

**Pull request merged**